source/libraries/ContractDeployer.ts

```diff
@@ -193,8 +196,9 @@ export class ContractDeployer {
     private async createGenesisUniverse(): Promise<ContractBlockchainData> {
         const delegatorBuilder = this.ethjsContract(this.signatures["Delegator"], this.bytecodes["Delegator"], { from: this.testAccounts[0], gas: this.gasAmount });
         const universeBuilder = this.ethjsContract(this.signatures["Universe"], this.bytecodes["Universe"], { from: this.testAccounts[0], gas: this.gasAmount });
-        const receiptAddress = await delegatorBuilder.new(this.controller.address, `0x${binascii.hexlify("Universe")}`);
-        const receipt = await this.ethjsQuery.getTransactionReceipt(receiptAddress);
+        const transactionHash = await delegatorBuilder.new(this.controller.address, `0x${binascii.hexlify("Universe")}`);
+        await waitForTransactionToBeSealed(this.ethjsQuery, transactionHash);
```

Comment:
It feels like this is the wrong place to sleep, in particular, we aren't doing an `eth_call` next, we are calling another on-chain transaction (`universe.initialize`).  Perhaps I am misunderstanding the need for the sleep here?

Reply:
When I was testing in Parity, sleeping just before creating the new market seemed to be sufficient.  While this may be working, I agree that it should be changed.  I didn't really have a way to test it with Geth until yesterday, when Scott helped me get set up using Geth.  I now have a way to debug better with Geth, so I'll try moving it to see if the error still occurs.

Reply:
Without the sleep here, the script does not wait until the transaction is sealed.  As I understand it `await myContract.new()` is equivalent to ethrpc's `onSent` callback, i.e. it just waits until the transaction is _broadcast_.  If there's _nothing_ downstream of this contract creation that requires this transaction to be sealed prior to proceeding then that is fine.  However, if anything downstream uses this contract for anything, you have to make sure it's sealed before moving on.

Reply:
Agreed.  My line of thinking is that in order to keep the deployment fast, even against a live chain with 15 second blocks, we should only be waiting for receipts when its actually necessary to make progress.  If we wait for a block to be mined after every transaction deployment will take over 30-60 minutes on a live Byzantium chain vs 1-2 minutes.

Reply:
Aside from the controller, the contracts are _uploaded_ at the same time, aren't they?  On geth (Ropsten) what appears to happen is all of the contract creation transactions are submitted simultaneously, and then slowly are sealed (since they're too big to cram all into one block).

Reply:
They are, but we _should_ be waiting for them to seal before making any `eth_call`s that depend on them being available on the chain.  This is what I'm getting at in that we should target our "wait for X" to just when we actually need to wait for X, rather than liberally waiting for X for everything.

Reply:
As I understand it the sequence of events should be as follows:
1. controller contract creation broadcast
2. controller contract creation sealed
3. all other contract creations broadcast (in parallel)
4. all other contract creations sealed
5. contract initialization functions broadcast (in parallel (?) -- not sure if this is true or even should be true)
6. contract initialization functions sealed
7. integration tests (create market, etc.) which are currently not working (or semi-working, on parity instantseal)
Is this the _actual_ sequence of events occurring when we run the deployer, in its current form?  If so, this seems like the correct (although not optimized) sequence to me.  If not, then I agree it should be fixed.

Reply:
No, the required sequence is:
1. all contracts uploaded (parallel)
2. contracts added to whitelist (parallel)
3. all contracts initialized (parallel)
4. integration tests are run

There is no need to wait for sealing for steps 1-3 because we don't need to do any `eth_call`s as part of the process.  I believe instantiating the market is the first time we need to make an `eth_call`, since we need to get the address of the universe.  We could technically work around this, but probably not worth the effort.

Reply:
Note the above only works if the backing node will return transaction receipts for unsealed blocks (everything except Geth).  Without the ability to get the receipt for unfinished blocks, we either need to wait for sealing to get the address of the newly created contracts, or we need to calculate the contract addresses client side.

Reply:
Currently the script uploads the Controller before the other contracts.  The other contracts are uploaded in parallel.  Since you need all contract creations sealed before you run step 4, is there any benefit to not waiting while you're in the previous steps?  The rate limiting step here is the number of blocks needed (since the contract creations require many blocks total); getting to parts of steps 2/3 sooner shouldn't make a difference if you still need all of 1-3 to be done before 4.

Transaction receipts for unsealed blocks (with accompanying contract addresses) are subject to change until the block is sealed.  In my view, it doesn't make sense to rely on them.

> (everything except Geth)

Meaning Parity, cpp-ethereum, and pyethereum/pyethapp?  Have you guys tested in all these?  The only two clients that see serious use are Parity and geth.  They should be mostly compatible, but when they're not, all of our code _must_ consider geth to be the default, as it is the most widely used client.

Reply:
@tinybike The only testing I've done so far of the deployment script is with Parity/Geth.